### PR TITLE
[wip] streaming interceptor

### PR DESCRIPTION
**Pull request by dd-caleb**
For: https://github.com/DataDog/dd-trace-go/issues/272

This is an implementation of the streaming interceptor for gRPC. Much of the code was reused from the existing unary interceptors.

A few questions:

1. This approach generates a trace on every send/recv of the stream on both sides. Should we be concerned about generating too many traces?
1. Should the send/recv traces have the same tags as the existing rpc traces? It would not be hard to add this if we want it.
1. The client stream is kind of complicated because it returns before the stream is closed, which means we have to delay finishing the span. From reading the documentation in the source code I *think* waiting for `stream.Context.Done()` should signal that the span should finish, but I'm not 100% sure. We should probably do a more exhaustive test on this so we're not leaking goroutines.

If the number of traces and tagging makes sense, I will add more unit tests.

### CODE REVIEW

**Comment by mbyio** [comment]
Thanks for starting on this so quickly! Some context - we currently are using streaming RPCs just to split up large unary messages (eg. for a file upload) so we might prefer to have the entire duration of a stream show as a single trace. But I could see a use case for one trace per send/recv too, eg. for streaming notifications from one service to another. Unfortunately, I'm not sure what the best strategy is.

**Comment by dd-caleb** [comment]
@mbyio thank you for the feedback! The current implementation will trace both the main stream request and the messages sent back and forth. In this example there were 10 sends/recvs:

screen shot 2018-07-23 at 7 00 29 pm

One solution could be an interceptor option to change what is traced for the streaming interceptor (currently our only interceptor option for gRPC is the service name). The default would still be full tracing, but the send/recv trace could be disabled, or vice-versa the outer start-stream trace. (I think there are circumstances where streaming could be used for a very long time, as basically a long-lived connection, and a trace for the entire duration isn't particularly useful) This is not hard to add so I could look into it tomorrow.

**Comment by mbyio** [comment]
Oh awesome! Thanks!

**Comment by gbbr** [comment]
> One solution could be an interceptor option to change what is traced for the streaming interceptor (currently our only interceptor option for gRPC is the service name). The default would still be full tracing, but the send/recv trace could be disabled, or vice-versa the outer start-stream trace. (I think there are circumstances where streaming could be used for a very long time, as basically a long-lived connection, and a trace for the entire duration isn't particularly useful) This is not hard to add so I could look into it tomorrow.

I think this solution sounds like the way to go. Start full fledged and then be able to narrow it down or fine tune it based on preferences. This is one of those cases where having what Opentracing calls "logging" or Opencensus calls "annotations" would come in handy.

File: contrib/google.golang.org/grpc/client.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+				var err error
+				stream, err = streamer(ctx, desc, cc, method, opts...)
+				return err
+			})
+		if err != nil {
+			span.Finish(withStreamError(err))
+			return nil, err
+		}
+		go func() {
+			<-stream.Context().Done()
```
[the remark this answers is not in the document]
Comment (dd-caleb) [reply contrib/google.golang.org/grpc/client.go]: Looks like the gRPC code does:

			select {
			case <-cc.ctx.Done():
				cs.finish(ErrClientConnClosing)
			case <-ctx.Done():
				cs.finish(toRPCErr(ctx.Err()))
			}

And `finish` does:

	cs.cancel()

And the docs say:

	// To ensure resources are not leaked due to the stream returned, one of the following
	// actions must be performed:
	//      1. Call Close on the ClientConn.
	//      2. Cancel the context provided.
	//      3. Call RecvMsg until a non-nil error is returned. A protobuf-generated
	//         client-streaming RPC, for instance, might use the helper function
	//         CloseAndRecv (note that CloseSend does not Recv, therefore is not
	//         guaranteed to release all resources).
	//      4. Receive a non-nil, non-io.EOF error from Header or SendMsg.

So I *think* we will only leak if gRPC was already going to leak.

File: contrib/google.golang.org/grpc/option.go [hunk 2]
```diff
@@ -1,13 +1,45 @@
 package grpc
 
-type interceptorConfig struct{ serviceName string }
+// The StreamMode is used to control what is traced in the
+// StreamClientInterceptor and StreamServerInterceptor
+type StreamMode byte
```
Comment (gbbr) [inline contrib/google.golang.org/grpc/option.go]: While I really love this pattern, I feel it's overkill here. I find it works well when there are many options and many combinations of them, while in this case we have only two:

* Trace calls
* Trace messages

The third one is implied when both are set. I'd dumb it down completely from an API standpoint, getting rid of bitwise operations and simply having options such as `grpctrace.WithStreamCalls` and `grpctrace.WithStreamMessages`. Much simpler IMO, what do you think?
